mlir/lib/Dialect/MemRef/Transforms/ResolveShapedTypeResultDims.cpp

```diff
@@ -94,6 +94,9 @@ struct DimOfReifyRankedShapedTypeOpInterface : public OpRewritePattern<OpTy> {
                                  reifiedResultShapes)))
       return failure();
     unsigned resultNumber = dimValue.getResultNumber();
+    // Do not apply pattern if the IR is invalid (dim out of bounds).
+    if (*dimIndex >= reifiedResultShapes[resultNumber].size())
+      return failure();
```

Comment:
can we systematically add a notifyMatchFailure for a better debugging experience of why patterns fail to apply?